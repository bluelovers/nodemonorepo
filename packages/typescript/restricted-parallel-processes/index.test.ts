import subject from './index'

const params = [
  'echo seg0 no0',
  'echo seg0 no1',
  'echo seg0 no2',
  'echo seg0 no3',
  'echo seg1 no0',
  'echo seg1 no1',
  'echo seg1 no2',
  'echo seg1 no3',
  'echo seg2 no0',
  'echo seg2 no1',
  'echo seg2 no2',
  'echo seg2 no3',
  'echo seg3 no0', // remaining
  'echo seg3 no1'
]

it('async iterator matches snapshot', async () => {
  const list = Array<ReadonlyArray<subject.ResultItem>>()

  for await (const item of subject(params, 4)) {
    list.push(item)
  }

  const sample = sortSegment(list)
  expect(sample).toMatchSnapshot()
})

it('array matches snapshot', async () => {
  const sample = sortSegment(await subject.asArray(params, 4))
  expect(sample).toMatchSnapshot()
})

it('calls oncreate', async () => {
  const allEventHandlers = Array<any>()
  const allCommandOutputs = Array<CommandOutput>()

  interface CommandOutput {
    readonly command: string
    readonly stdout: string
    readonly status: number
    readonly signal: string | null
  }

  const oncreate: subject.ObjectParams.CreationEventHandler = event => {
    allEventHandlers.push(event.params.oncreate)

    let stdout = ''

    // @ts-ignore
    event.process.stdout.on('data', chunk => {
      stdout += chunk
    })

    event.process.on('close', (status, signal) => {
      allCommandOutputs.push({
        command: [event.params.command, ...event.params.argv || []].join(' '),
        stdout: stdout.trim(),
        status,
        signal
      })
    })
  }

  const objparams = params
    .map(subject.getObjectParam)
    .map(x => ({ ...x, oncreate }))

  await subject.asArray(objparams, 4)

  expect(allEventHandlers).toEqual(allEventHandlers.map(() => oncreate))
  expect(allCommandOutputs.sort((a, b) => a.command > b.command ? 1 : -1)).toMatchSnapshot()
})

function getStringFromChunks (chunks: subject.ResultItem.DataChunkList): string {
  return chunks.map(x => String(x)).join('').trim()
}

function convertResultItem (item: subject.ResultItem) {
  class Result {
    readonly status = item.status
    readonly signal = item.signal
    readonly stdout = getStringFromChunks(item.stdout)
    readonly stderr = getStringFromChunks(item.stderr)
  }

  return new Result()
}

function sortSegment (item: subject.asArray.ReturningValue) {
  return item.map(
    part => part
      .map(convertResultItem)
      .sort((a, b) => a.stdout > b.stdout ? 1 : -1)
  )
}
