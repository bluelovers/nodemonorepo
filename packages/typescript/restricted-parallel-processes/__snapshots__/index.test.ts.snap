// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`array matches snapshot 1`] = `
Array [
  Array [
    Result {
      "signal": null,
      "status": 0,
      "stderr": "",
      "stdout": "seg0 no0",
    },
    Result {
      "signal": null,
      "status": 0,
      "stderr": "",
      "stdout": "seg0 no1",
    },
    Result {
      "signal": null,
      "status": 0,
      "stderr": "",
      "stdout": "seg0 no2",
    },
    Result {
      "signal": null,
      "status": 0,
      "stderr": "",
      "stdout": "seg0 no3",
    },
  ],
  Array [
    Result {
      "signal": null,
      "status": 0,
      "stderr": "",
      "stdout": "seg1 no0",
    },
    Result {
      "signal": null,
      "status": 0,
      "stderr": "",
      "stdout": "seg1 no1",
    },
    Result {
      "signal": null,
      "status": 0,
      "stderr": "",
      "stdout": "seg1 no2",
    },
    Result {
      "signal": null,
      "status": 0,
      "stderr": "",
      "stdout": "seg1 no3",
    },
  ],
  Array [
    Result {
      "signal": null,
      "status": 0,
      "stderr": "",
      "stdout": "seg2 no0",
    },
    Result {
      "signal": null,
      "status": 0,
      "stderr": "",
      "stdout": "seg2 no1",
    },
    Result {
      "signal": null,
      "status": 0,
      "stderr": "",
      "stdout": "seg2 no2",
    },
    Result {
      "signal": null,
      "status": 0,
      "stderr": "",
      "stdout": "seg2 no3",
    },
  ],
  Array [
    Result {
      "signal": null,
      "status": 0,
      "stderr": "",
      "stdout": "seg3 no0",
    },
    Result {
      "signal": null,
      "status": 0,
      "stderr": "",
      "stdout": "seg3 no1",
    },
  ],
]
`;

exports[`async iterator matches snapshot 1`] = `
Array [
  Array [
    Result {
      "signal": null,
      "status": 0,
      "stderr": "",
      "stdout": "seg0 no0",
    },
    Result {
      "signal": null,
      "status": 0,
      "stderr": "",
      "stdout": "seg0 no1",
    },
    Result {
      "signal": null,
      "status": 0,
      "stderr": "",
      "stdout": "seg0 no2",
    },
    Result {
      "signal": null,
      "status": 0,
      "stderr": "",
      "stdout": "seg0 no3",
    },
  ],
  Array [
    Result {
      "signal": null,
      "status": 0,
      "stderr": "",
      "stdout": "seg1 no0",
    },
    Result {
      "signal": null,
      "status": 0,
      "stderr": "",
      "stdout": "seg1 no1",
    },
    Result {
      "signal": null,
      "status": 0,
      "stderr": "",
      "stdout": "seg1 no2",
    },
    Result {
      "signal": null,
      "status": 0,
      "stderr": "",
      "stdout": "seg1 no3",
    },
  ],
  Array [
    Result {
      "signal": null,
      "status": 0,
      "stderr": "",
      "stdout": "seg2 no0",
    },
    Result {
      "signal": null,
      "status": 0,
      "stderr": "",
      "stdout": "seg2 no1",
    },
    Result {
      "signal": null,
      "status": 0,
      "stderr": "",
      "stdout": "seg2 no2",
    },
    Result {
      "signal": null,
      "status": 0,
      "stderr": "",
      "stdout": "seg2 no3",
    },
  ],
  Array [
    Result {
      "signal": null,
      "status": 0,
      "stderr": "",
      "stdout": "seg3 no0",
    },
    Result {
      "signal": null,
      "status": 0,
      "stderr": "",
      "stdout": "seg3 no1",
    },
  ],
]
`;

exports[`calls oncreate 1`] = `
Array [
  Object {
    "command": "echo seg0 no0",
    "signal": null,
    "status": 0,
    "stdout": "seg0 no0",
  },
  Object {
    "command": "echo seg0 no1",
    "signal": null,
    "status": 0,
    "stdout": "seg0 no1",
  },
  Object {
    "command": "echo seg0 no2",
    "signal": null,
    "status": 0,
    "stdout": "seg0 no2",
  },
  Object {
    "command": "echo seg0 no3",
    "signal": null,
    "status": 0,
    "stdout": "seg0 no3",
  },
  Object {
    "command": "echo seg1 no0",
    "signal": null,
    "status": 0,
    "stdout": "seg1 no0",
  },
  Object {
    "command": "echo seg1 no1",
    "signal": null,
    "status": 0,
    "stdout": "seg1 no1",
  },
  Object {
    "command": "echo seg1 no2",
    "signal": null,
    "status": 0,
    "stdout": "seg1 no2",
  },
  Object {
    "command": "echo seg1 no3",
    "signal": null,
    "status": 0,
    "stdout": "seg1 no3",
  },
  Object {
    "command": "echo seg2 no0",
    "signal": null,
    "status": 0,
    "stdout": "seg2 no0",
  },
  Object {
    "command": "echo seg2 no1",
    "signal": null,
    "status": 0,
    "stdout": "seg2 no1",
  },
  Object {
    "command": "echo seg2 no2",
    "signal": null,
    "status": 0,
    "stdout": "seg2 no2",
  },
  Object {
    "command": "echo seg2 no3",
    "signal": null,
    "status": 0,
    "stdout": "seg2 no3",
  },
  Object {
    "command": "echo seg3 no0",
    "signal": null,
    "status": 0,
    "stdout": "seg3 no0",
  },
  Object {
    "command": "echo seg3 no1",
    "signal": null,
    "status": 0,
    "stdout": "seg3 no1",
  },
]
`;
